includes/Story_Renderer/HTML.php

```diff
@@ -111,143 +73,100 @@ public function __construct( Story $story ) {
 	 * @return string The complete HTML markup for the story.
 	 */
 	public function render() {
-		$markup = self::DEFAULT_DOCTYPE . $this->story->get_markup();
-		$markup = $this->adapt_encoding( $markup );
+		$markup = $this->story->get_markup();
 		$markup = $this->replace_html_head( $markup );
-		$markup = $this->remove_noscript_amp_boilerplate( $markup );
 
-		$this->document = $this->load_html( $markup );
+		// If the AMP plugin is installed and available in a version >= than ours,
+		// all sanitization and optimization should be delegated to the AMP plugin.
+		if ( defined( '\AMP__VERSION' ) && version_compare( AMP__VERSION, WEBSTORIES_AMP_VERSION, '>=' ) ) {
+			add_filter( 'amp_content_sanitizers', [ $this, 'add_amp_content_sanitizers' ] );
 
-		// Run all further transformations on the DOMDocument.
-
-		$this->add_noscript_amp_boilerplate();
-		$this->transform_html_start_tag();
-		$this->transform_a_tags();
-		$this->insert_analytics_configuration();
-
-		$this->add_poster_images();
-		$this->add_publisher_logo();
-
-		return trim( (string) $this->document->saveHTML() );
-	}
-
-	/**
-	 * Adapt the encoding of the content.
-	 *
-	 * @link https://github.com/ampproject/amp-wp/blob/a393acf701e8e44d80225affed99d528ee751cb9/lib/common/src/Dom/Document.php
-	 *
-	 * @since 1.0.0
-	 *
-	 * @param string $markup Source content to adapt the encoding of.
-	 *
-	 * @return string Adapted content.
-	 */
-	protected function adapt_encoding( $markup ) {
-		$encoding = self::UNKNOWN_ENCODING;
-
-		if ( function_exists( 'mb_detect_encoding' ) ) {
-			$encoding = mb_detect_encoding( $markup, self::ENCODING_DETECTION_ORDER, true );
+			return $markup;
 		}
 
-		if ( function_exists( 'mb_convert_encoding' ) ) {
-			$markup = mb_convert_encoding( $markup, 'HTML-ENTITIES', $encoding );
-		}
+		$document = Document::fromHtml( $markup, get_bloginfo( 'charset' ) );
 
-		return $markup;
-	}
+		// This  should never actually happen.
+		if ( ! $document ) {
+			ob_start();
+			wp_die(
+				esc_html__( 'There was an error generating the web story, probably because of a server misconfiguration. Try contacting your hosting provider or open a new support request.', 'web-stories' ),
+				esc_html__( 'Web Stories', 'web-stories' ),
+				[
+					'response'  => 500,
+					'link_url'  => esc_url( __( 'https://wordpress.org/support/plugin/web-stories/', 'web-stories' ) ),
+					'link_text' => esc_html__( 'Visit Support Forums', 'web-stories' ),
+					'exit'      => false,
```

Comment:
@swissspidy I see :)

```suggestion
					'exit'      => true,
```

Reply:
Added this on purpose so I can write tests :)

Reply:
In this way, `wp_die()` is just being used to generate the error screen.

Reply:
That is all right.
A terminating function not terminating is considered "not testable code".
So benefiting from static analysis costs overriding the default handlers.

Reply:
I still don't see how this is so dissimilar from let's say `get_posts()` which has a different return type depending on its arguments 🤷‍♂️ 
That's why I was hoping for a way to dynamically tell phpstan whether the function is terminating or not.

Reply:
No, Pascal, it not different.

Understanding `get_posts()` takes this amount of code: https://github.com/szepeviktor/phpstan-wordpress/blob/master/src/GetPostsDynamicFunctionReturnTypeExtension.php

A proper terminating function takes 1 line of configuration. That is why I say: avoiding `exit: false` is sustainable.